modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1943 @@
+        if (clusterSnpReq != null)
+            return true;
+
+        synchronized (snpOpMux) {
+            return clusterSnpReq != null || clusterSnpFut != null;
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<String> getSnapshots() {
+        if (cctx.kernalContext().clientNode())
+            throw new UnsupportedOperationException("Client and daemon nodes can not perform this operation.");
+
+        synchronized (snpOpMux) {
+            return Arrays.stream(locSnpDir.listFiles(File::isDirectory))
+                .map(File::getName)
+                .collect(Collectors.toList());
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteFuture<Void> createSnapshot(String name) {
+        A.notNullOrEmpty(name, "name");
+
+        try {
+            if (cctx.kernalContext().clientNode())
+                throw new UnsupportedOperationException("Client and daemon nodes can not perform this operation.");
+
+            if (!IgniteFeatures.allNodesSupports(cctx.discovery().allNodes(), PERSISTENCE_CACHE_SNAPSHOT))
+                throw new IgniteException("Not all nodes in the cluster support a snapshot operation.");
+
+            if (!active(cctx.kernalContext().state().clusterState().state()))
+                throw new IgniteException("Snapshot operation has been rejected. The cluster is inactive.");
+
+            DiscoveryDataClusterState clusterState = cctx.kernalContext().state().clusterState();
+
+            if (!clusterState.hasBaselineTopology())
+                throw new IgniteException("Snapshot operation has been rejected. The baseline topology is not configured for cluster.");
+
+            ClusterSnapshotFuture snpFut0;
+
+            synchronized (snpOpMux) {
+                if (clusterSnpFut != null && !clusterSnpFut.isDone())
+                    throw new IgniteException("Create snapshot request has been rejected. The previous snapshot operation was not completed.");
+
+                if (clusterSnpReq != null)
+                    throw new IgniteException("Create snapshot request has been rejected. Parallel snapshot processes are not allowed.");
+
+                if (getSnapshots().contains(name))
+                    throw new IgniteException("Create snapshot request has been rejected. Snapshot with given name already exists.");
+
+                snpFut0 = new ClusterSnapshotFuture(UUID.randomUUID(), name);
+
+                clusterSnpFut = snpFut0;
+                lastSeenSnpFut = snpFut0;
+            }
+
+            List<Integer> grps = cctx.cache().persistentGroups().stream()
+                .filter(g -> cctx.cache().cacheType(g.cacheOrGroupName()) == CacheType.USER)
+                .filter(g -> !g.config().isEncryptionEnabled())
+                .map(CacheGroupDescriptor::groupId)
+                .collect(Collectors.toList());
+
+            List<ClusterNode> srvNodes = cctx.discovery().serverNodes(AffinityTopologyVersion.NONE);
```

Comment:
At some point during the snapshot creation, we need to check the partitions availability (offline baseline nodes, LOST partitions) and prohibit snapshot creation if at least one partition is unavailable.

Reply:
Alexey, these guarantees currently we've implemented it this PR:
1. If the affected BLT node left the cluster the snapshot process will be interrupted, intermediate changes will be deleted. Users will get an error.
2. Cache stop request currently blocked if a snapshot operation `in-progress`.
3. We are adding to snapshot only `OWNING` partitions if some of the partitions missed users will see `warn` message in the log which partitions (index) were missed.
4. If for of requested cache groups `CacheGroupContext` is null the snapshot operation will be interrupted.
5. If partition files are not accessible the snapshot operation will be interrupted and the user will see an error.

Please let me know if I'm missing something.

Reply:
I was thinking about the following case: 4 nodes, 1 backup. Say, partition 0 is located on nodes A and B. Nodes A and B stop before snapshot is initiated. At this point we will have partition either on LOST state, or we will not have this partition in partitions map. If the snapshot is triggered after A and B went down, the snapshot creation will not hit any of the conditions you've mentioned: no nodes will leave the cluster during the snapshot, there will be no partitions in non-OWNING state (the rebalance will not start because baseline did not change).

Reply:
In my view, a snapshot procedure just `freezes` the current cluster state. If the cluster has `LOST` partitions (not all baseline nodes are online) why we should interrupt the snapshot procedure? The process creates a cluster snapshot `as-is`.

Your case is mostly related to the recovery procedure. The additional information for the created snapshot will be added further and, for example, in snapshot-view users will see that the snapshot has `LOST` partitions and it's the point of decision - will he start recovery from this snapshot or not. 

Any way we can add a warning message. 
WDYT?